### PR TITLE
look up users and sites by id or name

### PR DESCRIPTION
**Pull request by ryanfkeepers**
Adds a lookup step to graph connector to find
an owner's id and name given some identifier.
The identifier, for either sites or users, can be a
well formed id or name.

---

#### Does this PR need a docs update or release note?

- [x] :no_entry: No

#### Type of change

- [x] :sunflower: Feature

#### Issue(s)

* #2825

#### Test Plan

- [x] :zap: Unit test
- [x] :green_heart: E2E

### CODE REVIEW

**Comment by ryanfkeepers** [comment]
@ashmrtn this is ready for review, if you would be so kind.

File: src/internal/connector/discovery/api/sites.go [hunk 1]
```diff
@@ -85,15 +87,67 @@ func (c Sites) GetAll(ctx context.Context, errs *fault.Bus) ([]models.Siteable,
...
+		err  error
+	)
+
+	ctx = clues.Add(ctx, "given_site_id", id)
+
+	if siteIDRE.MatchString(id) {
+		resp, err = c.stable.Client().SitesById(id).Get(ctx, nil)
+		if err != nil {
+			return nil, graph.Wrap(ctx, err, "getting site by id")
+		}
```
Comment (ashmrtn) [inline src/internal/connector/discovery/api/sites.go:114]: I think we can just exit early from this and not require an `else` block?
```suggestion
		}
		
		return resp, err
```

File: src/internal/connector/discovery/api/sites.go [hunk 2]
```diff
@@ -85,15 +87,67 @@ func (c Sites) GetAll(ctx context.Context, errs *fault.Bus) ([]models.Siteable,
...
+		var (
+			url   = strings.TrimPrefix(id, "https://")
+			parts = strings.SplitN(url, "/", 1)
+			host  = parts[0]
+			path  string
+		)
+
+		if len(parts) > 1 {
+			path = parts[1]
+		}
```
Comment (ashmrtn) [inline src/internal/connector/discovery/api/sites.go:125]: this should probably just use the URL package in golang so we don't have to worry about parsing it ourselves and other weirdness that can cause

parsing URLs: https://pkg.go.dev/net/url#Parse
info in a URL: https://pkg.go.dev/net/url#URL

File: src/internal/connector/graph_connector.go [hunk 3]
```diff
@@ -223,3 +175,139 @@ func (gc *GraphConnector) incrementAwaitingMessages() {
...
+
+	var (
+		id, name string
+		err      error
+	)
+
+	// if r.enum == Sites {
+	// TODO: check all suffixes in nameToID
+	// }
+
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/connector/graph_connector.go:265]: Nope, that gets built into the next PR.

File: src/internal/connector/graph_connector.go [hunk 4]
```diff
@@ -223,3 +175,139 @@ func (gc *GraphConnector) incrementAwaitingMessages() {
...
+var ErrResourceOwnerNotFound = clues.New("resource owner not found in tenant")
+
+// getOwnerIDAndNameFrom looks up the owner's canonical id and display name.
+// if idToName and nameToID are populated, and the owner is a key of one of
+// those maps, then those values are returned.  As a fallback, the resource
+// calls the discovery api to fetch the user or site using the owner value.
+// This fallback assumes that the owner is a well formed ID or display name
+// of appropriate design (PrincipalName for users, WebURL for sites).
+// If the fallback lookup is used, the maps are populated to contain the
+// id and name references.
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/connector/graph_connector.go:240]: Ah, outdated comments, thanks.

File: src/internal/connector/graph_connector.go [hunk 5]
```diff
@@ -223,3 +175,139 @@ func (gc *GraphConnector) incrementAwaitingMessages() {
...
+// The id-name swapper is optional.  Some processes will look up all owners in
+// the tenant before reaching this step.  In that case, the data gets handed
+// down for this func to consume instead of performing further queries.  The
+// maps get stored inside the gc instance for later re-use.
+//
+// TODO: If the maps are nil or empty, this func will perform a lookup on the given
+// owner, and populate each map with that owner's id and name for downstream
+// guarantees about that data being present.  Optional performance enhancement
+// idea: downstream from here, we should _only_ need the given user's id and name,
+// and could store minimal map copies with that info instead of the whole tenant.
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/connector/graph_connector.go:291]: I'll delete it.

File: src/internal/connector/graph_connector_test.go [hunk 6]
```diff
@@ -39,122 +39,216 @@ func TestGraphConnectorUnitSuite(t *testing.T) {
...
+) (string, string, error) {
+	return mnig.id, mnig.name, nil
+}
+
 func (suite *GraphConnectorUnitSuite) TestPopulateOwnerIDAndNamesFrom() {
 	const (
-		ownerID   = "owner-id"
-		ownerName = "owner-name"
+		id   = "owner-id"
+		name = "owner-name"
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/connector/graph_connector_test.go:58]: haha, he can be trained!

File: src/internal/connector/discovery/api/sites.go (hunk 2 above)
Comment (ryanfkeepers) [reply src/internal/connector/discovery/api/sites.go]: Huh, the funny thing is I have a different PR where I did just that.  Wonder why my brain skipped out on me here?

**Comment by ryanfkeepers** [comment]
@ashmrtn ready for the next review when you are, please and thank you!

File: src/internal/connector/discovery/api/sites.go [hunk 7]
```diff
@@ -85,15 +88,74 @@ func (c Sites) GetAll(ctx context.Context, errs *fault.Bus) ([]models.Siteable,
...
+// 10rqc2.sharepoint.com,deadbeef-0000-0000-0000-000000000000,beefdead-0000-0000-0000-000000000000
+// deadbeef-0000-0000-0000-000000000000,beefdead-0000-0000-0000-000000000000
+var siteIDRE = regexp.MustCompile(`(.+,)?` + uuidRE + "," + uuidRE)
+
+const webURLGetTemplate = "https://graph.microsoft.com/v1.0/sites/%s:/%s"
+
+// GetByID looks up the site matching the given ID.  The ID can be either a
+// canonical site id or a webURL.  Assumes the webURL is complete and well formed;
+// eg: https://10rqc2.sharepoint.com/sites/Example
 func (c Sites) GetByID(ctx context.Context, id string) (models.Siteable, error) {
```
Comment (ashmrtn) [inline src/internal/connector/discovery/api/sites.go:103]: this is bike shedding a bit, but I see "ID" and think like M365 ID. But this takes either a URL or M365 ID so it's a little confusing. If there's something else we could use besides ID that may be helpful

File: src/internal/connector/discovery/api/sites.go [hunk 8]
```diff
@@ -116,23 +178,23 @@ func validateSite(item any) (models.Siteable, error) {
...
+	if len(wURL) == 0 {
 		return nil, clues.New("missing webURL").With("site_id", id) // TODO: pii
 	}
 
 	// personal (ie: oneDrive) sites have to be filtered out server-side.
-	if strings.Contains(url, personalSitePath) {
+	if strings.Contains(wURL, personalSitePath) {
 		return nil, clues.Stack(errKnownSkippableCase).
-			With("site_id", id, "site_url", url) // TODO: pii
+			With("site_id", id, "site_web_url", wURL) // TODO: pii
```
Comment (ashmrtn) [inline src/internal/connector/discovery/api/sites.go:189]: nit: this message and the one below are inconsistent for the web URL key (site_web_url vs site_url)

**Review by ashmrtn — APPROVED** [review]

**Pull request merged**